### ISSUE TITLE
Print a warning when relying on case insensitivity

### ISSUE THREAD

**Issue by jlnr** [gosu/gosu#363]
Everytime a file is loaded in Gosu, there should be a check whether the case matches or not. If it does not match, then a warning should be printed.

Hopefully this will solve the issue that games written on Windows and macOS often do not work on Linux because of subtle case differences in resource filenames.

**Comment by Kjarrigan** [comment]
Your intention is good but I fear its not that easy to implement. I just googled a bit but it seems really complicated and involve a lot of (platform-specific) code to get the filename/path in the correct case to match against the given one. I don't think that this little help would justify that many code (e.g. http://stackoverflow.com/questions/74451/getting-actual-file-name-with-proper-casing-on-windows).

**Comment by jlnr** [comment]
Hmmm. It seems to be hard to do in C, but what if we limit it to Ruby/Gosu? `Dir[]` seems to do all the heavy-lifting on macOS at least:
```
Dir["/librarY/prEFerences"]
=> ["/Library/Preferences"]
```

**Comment by Kjarrigan** [comment]
I thought about that too, but wasn't sure if thats an option for you :smile: I wonder - do you have any information about the total / relative numbers of Gosu users and which language they prefer - are there more C++ or more ruby Gosu projects. I guess through the RubyGems download statistics we can somewhat estimate the Ruby part, but is there any equal "counter" for C++? Just out of curiosity.

**Comment by jlnr** [comment]
No, there's no counter for C++ users. Given how few C++ Gosu projects or questions are posted on the boards, I would estimate that there are 10-100 Ruby users per C++ user.

**Comment by pmer** [comment]
Since Ruby is just going to be calling some C-based API under the covers (or Objective-C...) that can be brought in to Gosu's C++ core.

**Comment by pmer** [comment]
@jlnr: Which version of Ruby are you using?

Here's `ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]`

```
irb(main):001:0> Dir["/librarY/prEFerences"]
=> ["/librarY/prEFerences"]
```

**Comment by jlnr** [comment]
> ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-darwin15]

Latest `ruby` from Homebrew. Interesting that that's a new feature.

**Comment by Kjarrigan** [comment]
@pmer Bringing it in the core would be the optimal solution but as mentioned above there seems to be no easy way in doing it in C++ across all plattforms without a ton of code. But I'm by no means familiar enough with C++ to be 100% certain, so if you know a good approach please take over this issue and submit a PR. Your help would be appreciated. I every know and then just scroll through the issues and try to find "simple" tasks to improve Gosu (and my knowledge about the internals). This sounded easy at first but after googl'ing a bit I didn't find an easy (& short) solution.

**Comment by pmer** [comment]
To find the correct case, a process with Ruby calls case folding, Ruby has some `#ifdef`'d code that calls [getattrlist on macOS](https://github.com/ruby/ruby/blob/cad4591086b285fa5c8fb61f19692b9a521dcea7/dir.c#L1564) and [GetFileAttributesExW on Windows](https://github.com/ruby/ruby/blob/cad4591086b285fa5c8fb61f19692b9a521dcea7/dir.c#L1621).

The logic surrounding this case folding feature in Ruby is pretty interesting and IMHO is probably not worth replicating in Gosu C++.

Fortunately, I found a way to force Ruby to correct a path's case, even on macOS's build-in Ruby 2.0.0:

```ruby
Dir.glob("/librarY/prEFerences", File::FNM_CASEFOLD)
=> ["/Library/Preferences"]
```

At some point in the past few years the `CASEFOLD` options was turned on by default in Ruby. Since macOS is using an old build it's missing this default, but we can force it on by explicitly requesting it like this.

**Comment by pmer** [comment]
Good intuition @Kjarrigan. Looks like you were right. :)

**Comment by pmer** [comment]
Side note: `Dir.glob` will also return a different string from one it was passed if UTF-8 normalization takes place, which happens on HFS+. So a mismatch between the glob and the requested file name could mean either a case fold or a normalization...

**Comment by Kjarrigan** [comment]
> File::FNM_CASEFOLD

Never have seen or used this one. Good catch @pmer. This will certainly help implementing this feature in Ruby across all plattforms and even for older Ruby's (any chance you know when this option was introdcued?). :+1: 

> So a mismatch between the glob and the requested file name could mean either a case fold or a normalization...

Atm I don't know how to deal with this as I'm no Mac-User, but lets see how noisy this feature is after it's implemented. Maybe we should provide an option to turn this on/off as it really has a special purpose. I guess you usually have to check your game only once (per release) for this possible problem on case-sensitive target systems.

**Comment by pmer** [comment]
@Kjarrigan Looks like it was [introduced in 2001](https://github.com/ruby/ruby/commit/6aa71d4c800d11d9735007cf3b063e5ea2fc5941). 👍

**Comment by jlnr** [comment]
UTF-8 normalisation on macOS has bitten me so hard with Subversion. I wonder if one could build a game that works on macOS, but not on Linux, because the filename and the string in the source file use different normalisations?

Ruby 2.2 allows us to normalise both filenames before comparing them:

https://ruby-doc.org/stdlib-2.2.2/libdoc/unicode_normalize/rdoc/UnicodeNormalize.html

Adding this feature for Ruby 1.8+ on Windows and 2.2+ on macOS sounds good enough to me.

Related issues:

#359 - it would help to pre-compose characters in `Font`, so maybe I should just add a simple composition table to Gosu. It doesn't need to be perfect to be helpful.

#356 - adding this Ruby-only patch is a bit frustrating with the current SWIG structure because it requires manually monkey-patching every method that accepts a string filename. If the Ruby wrapper were hand-written, it could just include the check without monkey-patching (calling Ruby is simple enough through `rb_eval`).
